### PR TITLE
Change protection algorithm

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #12 #15

Add autogenerate token option for `CsrfToken`. 
Add token generator interface and simple implementation.
Add storage that usage generator `GeneratorCsrfTokenStorage`.

Review this concept, if ok then I add phpdoc and tests.

### CODE REVIEW

File: src/CsrfMiddleware.php [hunk 1]
```diff
@@ -24,23 +23,19 @@ final class CsrfMiddleware implements MiddlewareInterface
     private string $headerName = self::HEADER_NAME;
 
     private ResponseFactoryInterface $responseFactory;
-    private CsrfTokenStorageInterface $storage;
+    private CsrfToken $token;
 
     public function __construct(
         ResponseFactoryInterface $responseFactory,
-        CsrfTokenStorageInterface $storage
+        CsrfToken $token
```
Comment (samdark) [inline src/CsrfMiddleware.php:30]:
```suggestion
        CsrfTokenService $tokenService
```

I think we should be clear that `CsrfToken` is **not** the data transfer object holding the token value.

Comment (samdark) [reply src/CsrfMiddleware.php:30]: May be named `CsrfTokenManager` or something alike alternatively to "service".

File: src/CsrfToken.php [hunk 2]
```diff
@@ -5,16 +5,24 @@
...
 use LogicException;
+use Yiisoft\Csrf\TokenGenerator\CsrfTokenGeneratorInterface;
 use Yiisoft\Csrf\TokenStorage\CsrfTokenStorageInterface;
 use Yiisoft\Security\TokenMask;
 
 final class CsrfToken
 {
+    private CsrfTokenGeneratorInterface $generator;
     private CsrfTokenStorageInterface $storage;
+    private bool $autoGenerate;
```
Comment (samdark) [inline src/CsrfToken.php:16]: Is it used anywhere except tests? Why should I turn it off?

**Review by samdark — COMMENTED** [review]
Overall 👍

**Comment by vjik** [comment]
#17 is better

**Pull request closed without merging**